py/dml/c_backend.py

```diff
@@ -2258,10 +2307,118 @@ def generate_alloc_trait_vtables(node):
         if subnode.traits:
             for trait in subnode.traits.referenced:
                 name = subnode.traits.vtable_cname(trait)
-                out(f'{name} = MM_MALLOC(1, typeof(*{name}));\n')
+                out(f'{name} = MM_ZALLOC(1, typeof(*{name}));\n')
     out('}\n', preindent=-1)
     splitting_point()
 
+def generate_startup_calls(devnode):
+    start_function_definition('void _startup_calls(void)')
+    out('{\n', postindent=1)
+
+    def startup_calls(startup_methods):
+        by_dims = {}
+        for (method_data_tuple, dims) in startup_methods:
+            by_dims.setdefault(dims, []).append(method_data_tuple)
+        all_dims = sorted(by_dims)
+        prev_dims = ()
```

Comment:
You're right that the code is copy-pasted, but I already stripped out the loop-splitting parts from it. `prev_dims` is part of the mechanism that allow loops to be merged, i.e. what allows this:
```
group g1[i < 3][j < 5] {
  group c[k < 7] {
    independent startup method m() { ... }
  }
  independent startup method n() { ... }
}

group g2[i < 3] {
  independent startup method o();
}
```
to be codegenned as this:
```
for (int i0 = 0; i0 < 3; ++i0) {
  for (int i1 = 0; i1 < 5; ++i1) {
    for (int i2 = 0; i2 < 7; ++i2) {
      m(i0, i1, i2);
    }
    n(i0, i1);
  }
  o(i0);
}
```
instead of genning a whole set of nested loops for each individual independent startup. I can very much believe such boilerplate would add up with an increasing number of startup methods across unique nodes, and while it's unlikely that number will be very large, seeing how we plan to have startup methods in templates in dml-builtins, I'd rather preemptively have this optimization in place.

Reply:
There will be a massive number of things done through startup calls, but for non-memoized I expect that we'll mostly do this from top level, like
```
independent startup method _register_attributes() {
  foreach a in (each conf_attribute in this) {
    register_attribute(..., a);
  }
}
```
Also, *if* this is a preformance problem, then there are better ways to optimize, e.g. group methods by number of dimensions and do a single table driven loop for each number of dimensions. Tables with constants are cheap.